Add tracing to kw and introduce `kw profiler`

davidbtadokoro opened the pull request:
This PR adds:

- Tracing capabilities
- A feature to visualize/analyze tracing reports

At the moment, I will maintain this PR as a draft, because I don't feel that the code is good enough for project standards, and I still need to fully validate if it doesn't affect normal installation (without tracing)

Below, is a backlog of things that need/should be done.

- [x] Split first commit; Although it represents a single change, we could spin some commits out of it to facilitate the review.
- [x] FIx `shfmt` error for `tracing/code_excerpts/profiler_feature.sh`
- [x] Add tests and doc for `tracing/tracing.sh`
- [x] Add tests and doc for `src/profiler.sh`
- [x] Add man page for `kw profiler`
- [x] Add completions for `kw profiler`

Commit:
scripts: profiler: Add tool for profiling kw executions

[What]

A "profiler" is a broad concept, but, in this context, it means software
that profiles (describes in detail) the execution of a program (a kw
feature) based on a tracing report.

[Why]

As tracing is implemented in the project, a tool that allows us to
visualize and analyze tracing reports should be handy.

A dedicated tool implemented inside the project grants the possibility
of having many ways to visualize the same tracing report, and tinker it
for our needs.

[How]

Add `scripts/profiler.sh` tool to produce a discrete profile of the flow
of execution using a tracing report produced by kw.

The tracing report directory is passed as an argument to the tool and
there are two types of profiles that can be produced.

1. `./scripts/profiler.sh --full <tracing_dir>`: Display full flow of
   execution for each thread. The output follows the format:

Thread nr. 0
--> profiler_main
  .976 milliseconds
  --> parse_profiler_options
    1.836 milliseconds
    --> kw_parse
      1.766 milliseconds
    <-- kw_parse
    1.398 milliseconds
  <-- parse_profiler_options
  1.184 milliseconds
  --> process_csv_files_for_full_profile
    213.107 milliseconds
  <-- process_csv_files_for_full_profile
  1.174 milliseconds
  --> display_full_profile
    2.208 milliseconds
  <-- display_full_profile
  .941 milliseconds
<-- profiler_main

2. `./scripts/profiler.sh --summary <tracing_dir>`: Outputs a CSV of the
   summary profile of the execution, where each entry has the name of
   the function and the total time spent on it in nanoseconds (e.g.
   'profiler_main,123456789').

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on kw:
```diff
@@ -204,6 +204,13 @@ function kw()
         pomodoro_main "$@"
       )
       ;;
+    profiler)
+      (
+        include "${KW_LIB_DIR}/profiler.sh"
+
+        profiler_main "$@"
+      )
+      ;;
```
If you try my suggestion from the previous commit, you will probably need to rethink this part. I see two possible paths:
1. You add some sort of "guards" around this code that only include this if the `--enable-trace` is used.
2. You create a new tool named `profile` under a folder named `tools/profile/`

davidbtadokoro replied:
Done

davidbtadokoro commented on kw:
```diff
@@ -55,6 +55,19 @@ fi
 # Export external variables required by kworkflow
 export KWORKFLOW
+
+# Abilitate tracing for functions defined in the kworkflow project
```
Done

davidbtadokoro commented on kw:
```diff
@@ -55,6 +55,19 @@ fi
+  mkdir --parents "$KW_TRACING_DIR"
+  if [[ "$?" != 0 ]]; then
+    printf 'Error: could not create tracing directory at %s\n' "$KW_TRACING_DIR"
+    exit 1
```
Done

davidbtadokoro commented on kw:
```diff
@@ -70,6 +83,10 @@ function kw()
 
+  if [[ -n "$KW_TRACING_DIR" ]]; then
+    printf "entry,${action},%s\n" "$(date '+%s%N')" >> "${KW_TRACING_DIR}/${THREAD_NUMBER}.csv"
+  fi
```
Done by using `/dev/shm` instead.

davidbtadokoro commented on kw:
```diff
@@ -295,6 +312,10 @@ function kw()
   esac
+
+  if [[ -n "$KW_TRACING_DIR" ]]; then
+    printf "exit,${action},%s\n" "$(date '+%s%N')" >> "${KW_TRACING_DIR}/${THREAD_NUMBER}.csv"
```
Done

davidbtadokoro commented on src/lib/kw_include.sh:
```diff
@@ -30,5 +32,65 @@ function include()
+
+  base_file_path=$(realpath "$base_file_path")
+
+  while read -r line; do
```
Done

Commit:
tracing: tracing: Introduce lib file for enabling tracing

To make an efficient tracing of kw, we need to avoid adding any overhead
at runtime, as one of the tracing objectives is to generate a report
that can be used to analyze performance. In reality its impossible to
not add any overhead, but we can heavilly mitigate it by injecting
tracing specific code when installing/building the application.

In this sense, add a new library file to house all functions for
injecting tracing-specific code into the application. To correctly trace
the execution, we need to inject code in two places:

1) Into the main kw file (the executable `kw` in the project root)
2) Into all kw library files (almost all files inside `src`)

This change implements a function named `sync_main_kw_file_with_tracing`
that takes care of the point number 1 above.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
tracing: tracing: Add functions to inject code to library files

As mentioned in the previous commit, to trace the flow of execution of
kw we need to inject code into the main kw executable (that was done in
the previous commit) and into library files (almost all files in `src`).

With that in mind, add three functions `sync_kw_lib_files_with_tracing`,
`inject_lib_file_with_tracing` and `get_tracing_log_line` to
`tracing/tracing.sh` that, combined, take care of injecting code that
enables tracing into all necessary files and synchronizes these with kw
local library (usually `~/.local/lib/kw`).

The code injected is the one that actually logs every entry and exit of
a function defined in the kw project. In other words, for every file
injected with code, we have:

- The first command of every function is to log the entrance in it;
- Before every `return` and `exit` statement in a function, there is a
  command to log the exiting of it;
- Before the end of a function definition (closing curly braces), there
  is a command to log the exiting of it preserving the return value of
  the last command;
- For every command that launches a background execution - which spuns
  another thread of execution - there is setup and teardown code (before
  and after the command) to separate the log of different threads.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
setup: Introduce installation with tracing enabled

Following the sequence of the last commits, add installation of kw with
tracing enabled. To install this version, run

./setup.sh --install --enable-tracing

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented on tests/tracing/tracing_test.sh:
```diff
@@ -62,4 +62,91 @@ some_commit_code='"'"'to test if the injection was successful'"'"'
+  assert_equals_helper 'The file was not correctly injected' "$LINENO" "$expected" "$output"
+
+  output=$(inject_lib_file_with_tracing "${SAMPLES_DIR}/tracing/src/simple_return_and_exit.sh")
+  expected='function foo()'$'\n'
+  expected+='{'$'\n'
+  expected+='printf '"'"'entry,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='printf '"'"'return,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='return 0'$'\n'
+  expected+='# shellcheck disable=SC2317'$'\n'
+  expected+='printf '"'"'exit,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='exit 0'$'\n'
+  expected+='local _return_val="$?"'$'\n'
+  expected+='printf '"'"'return,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='return "$_return_val"'$'\n'
+  expected+='}'
```
Maybe try heredoc.

davidbtadokoro replied:
Done

rodrigosiqueira commented on tests/tracing/tracing_test.sh:
```diff
@@ -62,4 +62,91 @@ some_commit_code='"'"'to test if the injection was successful'"'"'
+  expected+='}'
+  assert_equals_helper 'The file was not correctly injected' "$LINENO" "$expected" "$output"
+
+  output=$(inject_lib_file_with_tracing "${SAMPLES_DIR}/tracing/src/alternate_return_and_exit.sh")
+  expected='function foo()'$'\n'
+  expected+='{'$'\n'
+  expected+='printf '"'"'entry,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='command1 && command2 && printf '"'"'return,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv" && return 0'$'\n'
+  expected+='command3 && command4 && printf '"'"'exit,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv" && exit 0'$'\n'
+  expected+='local _return_val="$?"'$'\n'
+  expected+='printf '"'"'return,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='return "$_return_val"'$'\n'
+  expected+='}'
+  assert_equals_helper 'The file was not correctly injected' "$LINENO" "$expected" "$output"
+
+  output=$(inject_lib_file_with_tracing "${SAMPLES_DIR}/tracing/src/background_execution.sh")
+  expected='function foo()'$'\n'
+  expected+='{'$'\n'
+  expected+='printf '"'"'entry,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='THREAD_NUMBER=$((++TOTAL_NUMBER_OF_THREADS))'$'\n'
+  expected+='command &'$'\n'
+  expected+='THREAD_NUMBER=0'$'\n'
+  expected+='local _return_val="$?"'$'\n'
+  expected+='printf '"'"'return,foo,%s\n'"'"' "$(date +%s%N)" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'$'\n'
+  expected+='return "$_return_val"'$'\n'
+  expected+='}'
+  assert_equals_helper 'The file was not correctly injected' "$LINENO" "$expected" "$output"
```
Split these other tests into their own test function.

davidbtadokoro replied:
Done

rodrigosiqueira commented on tracing/tracing.sh:
```diff
@@ -31,3 +31,145 @@ function sync_main_kw_file_with_tracing()
+  local all_filepaths
+  local tmp_dir_path
+
+  tmp_dir_path=$(mktemp --directory)
```
Is it possible that mktemp fails?

davidbtadokoro replied:
Thanks for the catch. However I couldn't think of a way that this specific `mktemp` command can fail. Differently from your other comment, where I setted a different base from the default (`mktemp --directory --tmp-dir=/dev/shm`), I could think of and reproduce fail cases.

rodrigosiqueira commented on tracing/tracing.sh:
```diff
@@ -31,3 +31,145 @@ function sync_main_kw_file_with_tracing()
+
+  while IFS=$'\n' read -r filepath; do
+    # We shouldn't try to inject code in the completions files nor the `VERSION` file
+    if [[ "$filepath" =~ src\/(_kw|bash_autocomplete\.sh|VERSION) ]]; then
```
Maybe you could create a global array named `ignored_files` where you add all file names that should be ignored and make this line more generic by checking the array. I also think this approach will make it easy to maintain the trace option in the future.

davidbtadokoro replied:
Done

rodrigosiqueira commented on tracing/tracing.sh:
```diff
@@ -31,3 +31,145 @@ function sync_main_kw_file_with_tracing()
+      file_with_tracing+="$line"$'\n'
+    fi
+
+    while [[ -n "$inside_function" ]]; do
```
Maybe you can create a dedicated function for the operations inside this loop; this can also enable you to add a lot of refined tests for this parser. Also, I think the switch/case will be more readable for this case.

davidbtadokoro replied:
Done

rodrigosiqueira commented on tracing/tracing.sh:
```diff
@@ -31,3 +31,145 @@ function sync_main_kw_file_with_tracing()
+  local expression
+
+  expression="printf '${tracing_event},${function_name},%s\n' "
+  expression+='"$(date '+%s%N')" >> "${TMP_TRACING_DIR}/${THREAD_NUMBER}.csv"'
```
Maybe we should consider adding a proper function for this date in the lib file?

davidbtadokoro replied:
Hmm, I see the point, but I'm not sure that this would be a better approach.

As this string `"$(date '+%s%N')"` isn't being evaluated in this function, but rather will be evaluated in any other function defined in `src`, if we substitute this string for a call to a function, say, `get_timestamp_in_nano_secs` defined in `src/lib/kw_date_and_time.sh` we would have to assure that every lib file included `kw_date_and_time.sh`.

Does my point makes sense?

rodrigosiqueira commented on src/lib/lore.sh:
```diff
@@ -423,8 +423,7 @@ function process_patchsets()
         # Process each patch in parallel
-        thread_for_process_patch "$PATCHSETS_PROCESSED" "$shared_dir_for_parallelism" \
-          "$processed_patchset" "$patch_url" "$patch_title" &
+        thread_for_process_patch "$PATCHSETS_PROCESSED" "$shared_dir_for_parallelism" "$processed_patchset" "$patch_url" "$patch_title" &
```
This solution sounds fragile... can you maybe expand your parser from the previous commit to check if the line ends with `\` and, based on that, compose with the following line?

davidbtadokoro replied:
Done

rodrigosiqueira commented on setup.sh:
```diff
@@ -323,13 +327,23 @@ function synchronize_files()
   mkdir -p "$binpath"
-  cmd_output_manager "cp $app_name $binpath" "$verbose"
-  ASSERT_IF_NOT_EQ_ZERO "The command 'cp $app_name $binpath' failed" "$?"
+  if [[ "$ENABLE_TRACING" == 1 ]]; then
+    sync_main_kw_file_with_tracing "$app_name" "$binpath" "$TRACING_CODE_EXCERPTS_DIR"
+    ASSERT_IF_NOT_EQ_ZERO 'Could not sync kw main file with tracing enabled' "$?"
+  else
+    cmd_output_manager "cp $app_name $binpath" "$verbose"
+    ASSERT_IF_NOT_EQ_ZERO "The command 'cp $app_name $binpath' failed" "$?"
+  fi
```
Nitpick: I guess the first `if` will usually be false; in this sense, it might be better to reverse the logic.

davidbtadokoro replied:
Done

rodrigosiqueira commented on setup.sh:
```diff
@@ -323,13 +327,23 @@ function synchronize_files()
   mkdir -p "$libdir"
-  cmd_output_manager "rsync -vr $SRCDIR/ $libdir" "$verbose"
-  ASSERT_IF_NOT_EQ_ZERO "The command 'rsync -vr $SRCDIR $libdir' failed" "$?"
+  if [[ "$ENABLE_TRACING" == 1 ]]; then
```
same as the previous comment.

davidbtadokoro replied:
Done

rodrigosiqueira commented on src/bash_autocomplete.sh:
```diff
@@ -15,8 +15,8 @@ function _kw_autocomplete()
                     maintainers kernel-config-manager config remote explore
-                    pomodoro report device backup debug mail env patch-hub
-                    clear-cache drm vm version man help'
+                    pomodoro profiler report device backup debug mail env
+                    patch-hub clear-cache drm vm version man help'
```
I feel that profiler as an option inside kw does not make a lot of sense... I mean, I can see a user thinking that a profiler is something they can use to profile a kernel performance. Tbh, I'm more inclined to make this profiler an external script... maybe something in a folder named `tools`;

davidbtadokoro replied:
Agreed. Done it, but in the already present `scripts` directory rather than creating a new directory. Chose this approach as `scripts/kwreview.sh` is also tool that isn't for kw end-user.

rodrigosiqueira commented on src/_kw:
I think you can squash this commit in the previous one.

davidbtadokoro replied:
Done

rodrigosiqueira commented on documentation/man/features/profiler.rst:
```diff
@@ -0,0 +1,43 @@
+==============
+kw-profiler
```
squash this commit in the one that introduces the profile.

davidbtadokoro replied:
Done

davidbtadokoro commented:
Updated it, but the update still a WIP.

davidbtadokoro commented on tests/samples/tracing/kw_main_file_mock:
```diff
@@ -0,0 +1,18 @@
+
+some_command "$arg1" "$arg2"
+
+#TRACING_SETUP
```
Done

davidbtadokoro commented on tests/tracing/tracing_test.sh:
```diff
@@ -0,0 +1,65 @@
+local some_commit_code
+
+some_commit_code='"'"'to test if the injection was successful'"'"'
+}'''
```
Done

davidbtadokoro commented on tracing/code_excerpts/tracing_setup:
```diff
@@ -0,0 +1,19 @@
+declare -g THREAD_NUMBER=0
+declare -g TRACING_SESSION_DIR
+TRACING_SESSION_DIR="${KW_TRACING_DIR}/$(date '+%Y-%m-%d_%H-%M-%S')"
+declare -g TMP_TRACING_DIR
```
Done

davidbtadokoro commented on tracing/code_excerpts/tracing_setup:
```diff
@@ -0,0 +1,19 @@
+fi
+
+# Create temporary directory in shared memory to reduce IO overhead of tracing
+TMP_TRACING_DIR=$(mktemp --directory --tmpdir=/dev/shm)
```
Done

davidbtadokoro commented on tracing/code_excerpts/tracing_setup:
```diff
@@ -0,0 +1,19 @@
+# Define directory to store tracing reports of kw executions
```
Done

davidbtadokoro commented on tracing/tracing.sh:
```diff
@@ -0,0 +1,33 @@
+# This function injects code that enables tracing into the main kw file and
+# installs this altered version.
+#
+# @kw_path: Name/Path to main kw file (we assume it resides in the current dir)
```
Done

davidbtadokoro commented on tracing/tracing.sh:
```diff
@@ -0,0 +1,33 @@
+#   into kw main file
+function sync_main_kw_file_with_tracing()
+{
+  local kw_path="$1"
+  local bin="$2"
```
Done

rodrigosiqueira commented:
Hi @davidbtadokoro , everything looks good!
To complete this PR, add a tutorial under the kw section. Maybe after this: https://kworkflow.org/content/developmentworkflow.html

Commit:
documentation: content: tracingandprofiling: Add documentation for tracing and profiling capabilities

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented:
Hey @rodrigosiqueira, thanks so much for the time spent on reviewing this monster PR 😬

Updated the PR adding the documentation. See if it's OK.

rodrigosiqueira commented:
Merged! Thanks a lot for this change.
